oi/type_graph/TopoSorter.cpp

```diff
@@ -85,30 +85,34 @@ namespace {
  * Other containers are not required to do this, but might still have this
  * behaviour.
  */
-bool containerAllowsIncompleteParams(const Container& c) {
+bool containerAllowsIncompleteParam(const Container& c, size_t i) {
   switch (c.containerInfo_.ctype) {
     case SEQ_TYPE:
     case LIST_TYPE:
     case UNIQ_PTR_TYPE:
     case SHRD_PTR_TYPE:
       // Also std::forward_list, if we ever support that
       // Would be good to have this as an option in the TOML files
-      return true;
+      if (i == 0)
+        return true;
+      return false;
     default:
       return false;
   }
 }
 }  // namespace
 
 void TopoSorter::visit(Container& c) {
-  if (!containerAllowsIncompleteParams(c)) {
-    for (const auto& param : c.templateParams) {
+  for (size_t i = 0; i < c.templateParams.size(); i++) {
+    if (!containerAllowsIncompleteParam(c, i)) {
+      const auto& param = c.templateParams[i];
       accept(param.type());
     }
   }
   sortedTypes_.push_back(c);
-  if (containerAllowsIncompleteParams(c)) {
-    for (const auto& param : c.templateParams) {
+  for (size_t i = 0; i < c.templateParams.size(); i++) {
+    if (containerAllowsIncompleteParam(c, i)) {
+      const auto& param = c.templateParams[i];
       acceptAfter(param.type());
```

Comment:
doesn't `acceptAfter` mean this only needs one loop? if both cases were merged into the first loop, `accept` would occur immediately but `acceptAfter` would be appropriately delayed. this way might be clearer to read though.